File: stlfed_py/ex.py
import requests
import datetime
import pandas as pd
import numpy as np
from statsmodels.tsa.holtwinters import ExponentialSmoothing
import matplotlib.pyplot as plt

# Shitty way around not asking for an API key, since returned data is raw text
url = 'https://fred.stlouisfed.org/graph/fredgraph.csv'
fred_id = 'MONAN' # MO Total Non-Farm Employment
start_date = '1800-01-01'
today = datetime.datetime.now().strftime('%Y-%m-%d')
end_date = today
r = requests.get(f'{url}?id={fred_id}&cosd={start_date}&coed={end_date}')

with open('fred_id.csv', 'w') as f:
    f.write(r.text)

df = pd.read_csv('fred_id.csv')
df[fred_id] = (df[fred_id] / 100)

df.index.freq = 'M'

# Fit HW model
model = ExponentialSmoothing(
    df[fred_id],
    trend = 'add',
    seasonal = 'add',
    seasonal_periods = 12
).fit()

# Get MAPE
pct_error = abs(df[fred_id] - model.fittedvalues) / df[fred_id]
mape = np.mean(pct_error)

# Forecast
h = 12
fcast = model.forecast(h)

# Full series
out = df[fred_id].append(fcast)

# Plotting, just for interactive use
plt.plot(df.index, df[fred_id], label = 'Actual')
plt.plot(range(len(df.index), len(df.index) + h), fcast, label = 'Fcast')
plt.legend(loc = 'best')
plt.show()

File: stlfed_R/R/ex.R
if (interactive()) {
    
    library(forecast)
    library(httr)
    library(lubridate)
    
    url <- 'https://fred.stlouisfed.org/graph/fredgraph.csv'
    fred_id <- 'MONAN' # MO Total Non-Farm Employment
    today <- Sys.Date()
    start_date <- today - years(5)
    end_date <- today
    
    # Shitty way around not asking for an API key
    # Should eventually call httr::GET() with a key, to directly return a dataframe
    tmpfile <- tempfile(fileext = '.csv')
    download.file(glue::glue('{url}?id={fred_id}&cosd={start_date}&coed={end_date}'), tmpfile)
    
    df_0 <- read.csv(tmpfile, strip.white = TRUE, stringsAsFactors = FALSE)
    
    # Prep
    df_0$DATE <- as_date(df_0$DATE)
    df_0$label <- 'Actual'
    
    # Fit HW model
    ts_0 <- ts(
        df_0[, fred_id],
        start = c(year(min(df_0$DATE)), month(min(df_0$DATE))),
        end = c(year(max(df_0$DATE)), month(max(df_0$DATE))),
        frequency = 12
    )
    model <- HoltWinters(ts_0, seasonal = 'additive')
    
    # Forecast
    h <- 12
    fcast <- forecast(model, h = h)
    
    # Get MAPE
    mape <- accuracy(fcast)[5]
    
    first_fcast_period <- max(df_0$DATE) + months(1)
    last_fcast_period <- max(df_0$DATE) + months(h)
    fcast_dates <- seq(first_fcast_period, last_fcast_period, by = 'month')
    
    df_fcast <- data.frame(
        DATE = fcast_dates,
        fcast$mean,
        MAPE = mape,
        stringsAsFactors = FALSE
    )
    # Can't easily parse names without metaprogramming, fuuuck that
    colnames(df_fcast)[2] <- fred_id
    
    # Plot, for interactive use
    plot(fcast)
    
}